File: hackathons/week2/board.py
#!/usr/bin/env python3

import random


def get_board(n=-1, distribution=None):
    """
    Generates a boggle board; 
    no arguments - with dimensions of 4x4, based on dice distribution of new boggle version
    arguments given - with dimensions of nxn, based on distribution obtained from with replacement sampling of list
    
    Arguments:
    n (optional) -- dimensions of board 
    distribution (optional) -- list of characters to be sampled from
    
    Returns:
    board -- 2d list of size n x n containing characters (where n=4 by default)
    """
    if(n==-1):
        die =  {0:'AAEEGN',
                1:"ELRTTY",
                2:"AOOTTW",
                3:"ABBJOO",
                4:"EHRTVW",
                5:"CIMOTU",
                6:"DISTTY",
                7:"EIOSST",
                8:"DELRVY",
                9:"ACHOPS",
                10:"HIMNQU",
                11:"EEINSU",
                12:"EEGHNW",
                13:"AFFKPS",
                14:"HLNNRZ",
                15:"DEILRX"}
        
        #initialise board
        board = [ ["" for j in range(4)] for i in range(4)]
        
        for i in range(16):
            row_no = i//4 #integer division
            col_no = i%4 
            dice_throw =  random.randint(0,5)
            board[row_no][col_no] = die[i][dice_throw]
    else:
        #initialise board
        board = [ ["" for j in range(n)] for i in range(n)]
        
        for i in range(n*n):
            row_no = i//n #integer division
            col_no = i%n 
            dice_throw =  random.randint(0,len(distribution)-1)
            board[row_no][col_no] = distribution[dice_throw]

    return board


def get_neighbours(idx, n):
    """
    Finds indices of neighbours of current index (including diagonals)

    Arguments:
    idx -- current index as 2-tuple representing row and column
    n -- dimension of corresponding square board

    Returns:
    neighbours -- list of indices of neighbouring tiles on board
    """
    neighbours = []
    row = idx[0]
    col = idx[1]

    adj_rows = [idx[0]]
    adj_cols = [idx[1]]

    n_range = set(range(n))

    if ((row - 1) in n_range):
        adj_rows.append(row - 1)
    if ((row + 1) in n_range):
        adj_rows.append(row + 1)
    if ((col - 1) in n_range):
        adj_cols.append(col - 1)
    if ((col + 1) in n_range):
        adj_cols.append(col + 1)

    for i in adj_rows:
        for j in adj_cols:
            neighbours.append((i, j))

    neighbours.remove(idx)
    return neighbours

File: hackathons/week6/board.js
'use strict';

class Board {
    /**
     * Constructor: initialize a board with given size and letter distribution
     * @param size: the board size
     * @param distribution: the array of characters to be sampled from
     */
    constructor(size, distribution) {
        this.size = size;
        this.distribution = distribution;
        this.board = [];
        this.initBoard();
    }

    /**
     * Return generated board
     * @return {Array}
     */
    get distriBoard() {
        return this.board;
    }

    /**
     * Return this board's size
     * @return {number} the board's size
     */
    get boardSize() {
        return this.board.length;
    }

    /**
     * Return neighbouring positions of the given position.
     * Navigating on the board is done horizontally, vertically and diagonally.
     * @param position: current position
     * @return {Array} an array of neighbouring tiles on the board
     */
    getNeighbours(position) {
        let neighbours = [];
        let row = position[0];
        let col = position[1];

        let adjRows = [row];
        let adjCols = [col];

        let sizeRange = Array.from(
            new Array(this.size),
            (val, index) => index
        );

        if (sizeRange.indexOf(row - 1) > -1) {
            adjRows.push(row - 1);
        }
        if (sizeRange.indexOf(row + 1) > -1) {
            adjRows.push(row + 1);
        }
        if (sizeRange.indexOf(col - 1) > -1) {
            adjCols.push(col - 1);
        }
        if (sizeRange.indexOf(col + 1) > -1) {
            adjCols.push(col + 1);
        }

        for (let i= 0; i < adjRows.length; i++) {
            for (let j = 0; j < adjCols.length; j++) {
                neighbours.push([adjRows[i], adjCols[j]]);
            }
        }

        // determine element equivalent to position in the neighbours array
        let ndx = -1;
        for (let i = 0; i < neighbours.length; i++) {
            let element = neighbours[i];
            if (element[0] === position[0] && element[1] === position[1]) {
                ndx = i;
                break;
            }
        }
        // remove element equivalent to position from the neighbours array
        if (ndx > -1) {
            neighbours.splice(ndx, 1);
        }
        return neighbours;
    }

    /**
     * Generate the board based on dice distribution from the new Boggle version.
     * The result is a 2 dimensional array/list containing randomly generated characters.
     */
    initBoard() {
        // initialize this.board with empty strings beforehand
        for (let x = 0; x < this.size; x++) {
            let rowArray = Array(this.size).fill("");
            this.board.push(rowArray);
        }

        let distributionSize = this.distribution[0].length;

        for (let i = 0; i < this.size * this.size; i++) {
            let row = Math.floor(i / this.size);  // integer division
            let col = i % this.size;
            let diceThrow = Math.floor(Math.random() * (distributionSize - 1));
            this.board[row][col] = this.distribution[i][diceThrow];
        }
    }
}